runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/client/grpc/GrpcDirectGetWorkStream.java

```diff
@@ -201,7 +198,7 @@ private void maybeSendRequestExtension(GetWorkBudget extension) {
   @Override
   protected synchronized void onNewStream() {
     workItemAssemblers.clear();
-    if (!isShutdown()) {
+    if (!hasReceivedShutdownSignal()) {
```

Comment:
this is racy since shutdown coudl occur right after this check. can we instead ensure that we handle stream close exception on the send below? or that we handle it at the call site to onNewStream?

Reply:
done